components/merkledb/src/validation.rs

```diff
@@ -1,37 +1,57 @@
 //! Validation helpers for index names.
 
-/// Validate index name.
-pub fn is_valid_index_name<S: AsRef<str>>(name: S) -> bool {
+use core::fmt;
+
+/// Validates index name.
+pub fn is_valid_index_full_name<S: AsRef<str>>(name: S) -> bool {
     name.as_ref()
         .as_bytes()
         .iter()
         .copied()
-        .all(is_allowed_latin1_char)
+        .all(|c| is_allowed_index_name_char(c) || c == 46)
 }
 
-/// Check that latin1 character is allowed in index name.
+/// Validates index name prefix, it shouldn't contain the dot.
+pub fn is_valid_index_name_component<S: AsRef<str>>(name: S) -> bool {
+    name.as_ref()
+        .as_bytes()
+        .iter()
+        .copied()
+        .all(is_allowed_index_name_char)
+}
+
+/// Checks that character is allowed in index name.
 /// Only these combination of symbols are allowed:
 ///
-/// `[0..9]`, `[a-z]`, `[A-Z]`, `_`, `-`, `.`
-pub fn is_allowed_latin1_char(c: u8) -> bool {
+/// `[0..9]`, `[a-z]`, `[A-Z]`, `_`, `-`
+pub fn is_allowed_index_name_char(c: u8) -> bool {
     match c {
-        48..=57   // 0..9
-        | 65..=90   // A..Z
-        | 97..=122  // a..z
-        | 45..=46   // -.
-        | 95        // _
-        => true,
+        b'0'..=b'9' | b'A'..=b'Z' | b'a'..=b'z' | b'-' | b'_' => true,
         _ => false,
     }
 }
 
-/// Calls the `is_valid_name` function with the given name and panics if it returns `false`.
-pub(crate) fn assert_valid_name<S: AsRef<str>>(name: S) {
+fn assert_valid_name<S: AsRef<str> + Copy + fmt::Debug, F>(name: S, predicate: F, desc: &str)
+where
+    F: Fn(S) -> bool,
+{
     if name.as_ref().is_empty() {
         panic!("Index name must not be empty")
     }
 
-    if !is_valid_index_name(name) {
-        panic!("Wrong characters using in name. Use: a-zA-Z0-9 and _-.");
+    if !predicate(name) {
+        panic!("Wrong characters using in name ({:?}). {}", name, desc);
     }
 }
+
+/// Calls the `is_valid_index_full_name` function with the given name
+/// and panics if it returns `false`.
+pub(crate) fn assert_index_valid_full_name<S: AsRef<str> + Copy + fmt::Debug>(name: S) {
+    assert_valid_name(name, is_valid_index_full_name, "Use: a-zA-Z0-9 and _-.")
+}
+
+/// Calls the `is_valid_index_name_component` function with the given
+/// name and panics if it returns `false`.
+pub(crate) fn assert_valid_index_name_component<S: AsRef<str> + Copy + fmt::Debug>(name: S) {
```

Comment:
Nit: I'd name this `assert_valid_name_component`, since the method isn't always used for indexes (e.g., it's used for `Group`).